spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/core/convert/DefaultDatastoreEntityConverter.java

```diff
@@ -67,39 +69,56 @@ public DefaultDatastoreEntityConverter(DatastoreMappingContext mappingContext,
 				new PersistentEntityParameterValueProvider<>(persistentEntity, propertyValueProvider, null);
 
 		EntityInstantiator instantiator = this.instantiators.getInstantiatorFor(persistentEntity);
-		R instance = instantiator.createInstance(persistentEntity, parameterValueProvider);
-		PersistentPropertyAccessor accessor = persistentEntity.getPropertyAccessor(instance);
-
-		persistentEntity.doWithProperties(
-				(PropertyHandler<DatastorePersistentProperty>) datastorePersistentProperty -> {
-					Object value = propertyValueProvider.getPropertyValue(datastorePersistentProperty);
-					accessor.setProperty(datastorePersistentProperty, value);
-				});
+		R instance;
+		try {
+			instance = instantiator.createInstance(persistentEntity, parameterValueProvider);
+			PersistentPropertyAccessor accessor = persistentEntity.getPropertyAccessor(instance);
+			persistentEntity.doWithProperties(
+					(PropertyHandler<DatastorePersistentProperty>) datastorePersistentProperty -> {
+						// if a property is a constructor argument, it was already computed on instantiation
+						if (!persistentEntity.isConstructorArgument(datastorePersistentProperty)) {
+							Object value = propertyValueProvider.getPropertyValue(datastorePersistentProperty);
+							accessor.setProperty(datastorePersistentProperty, value);
+						}
+					});
+		}
+		catch (DatastoreDataException e) {
+			throw new DatastoreDataException("Unable to read " + persistentEntity.getName() + " entity", e);
+		}
 
 		return instance;
 	}
 
 	@Override
+	@SuppressWarnings("unchecked")
 	public void write(Object source, Entity.Builder sink) {
 		DatastorePersistentEntity<?> persistentEntity = this.mappingContext.getPersistentEntity(source.getClass());
 		PersistentPropertyAccessor accessor = persistentEntity.getPropertyAccessor(source);
 		persistentEntity.doWithProperties(
-				(DatastorePersistentProperty datastorePersistentProperty) ->
-						writeProperty(sink, accessor, datastorePersistentProperty));
-	}
-
-	private void writeProperty(Entity.Builder sink, PersistentPropertyAccessor accessor,
-			DatastorePersistentProperty persistentProperty) {
-		Object propertyVal = accessor.getProperty(persistentProperty);
-
-		propertyVal = this.conversions.convertOnWrite(propertyVal);
-
-		Value val = DatastoreNativeTypes.wrapValue(propertyVal, persistentProperty);
-		if (persistentProperty.isUnindexed()) {
-			ValueBuilder valueBuilder = val.toBuilder();
-			valueBuilder.setExcludeFromIndexes(true);
-			val = valueBuilder.build();
-		}
-		sink.set(persistentProperty.getFieldName(), val);
+				(DatastorePersistentProperty persistentProperty) -> {
+					try {
+						Object val = accessor.getProperty(persistentProperty);
+						//Check if property is a non-null array
+						if (val != null && persistentProperty.isArray() && val.getClass() != byte[].class) {
```

Comment:
Since you've moved the list-vs-single conversion code to convertOnWrite, why not move this byte[] checking case in there as well?